### ISSUE TITLE
Can't get ROSA predictions on new data

### ISSUE THREAD

**Issue by skylogic004** [khliland/multiblock#2]
I'm getting an error when using `predict.rosa`  on a ROSA model. 
https://github.com/khliland/multiblock/blob/87e18bcf7d38e5b840830f63fbd51e2606e83770/R/rosa_results.R#L52

The error is `length of 'dimnames' [1] not equal to array extent`.

I narrowed it down to the call to `coef` where `intercept = TRUE` is passed:
```R
B <- coef(object, ncomp = ncomp, intercept = TRUE)
```

But, if I understand the ROSA model correctly, ROSA does not have an intercept? Passing `intercept = FALSE` makes the error go away but then the code that follows in `predict.rosa` also assumes that an intercept is present and calculates the predicted values using it.

I believe the correct code is something more like this:
```R
B <- coef(object, ncomp = ncomp, intercept = FALSE)
Bavg <- rowMeans(B, dims=2)
B0 <- object$Ymeans - object$Xmeans %*% B
pred <- newX %*% Bavg + rep(c(B0), each = nobs)
```
Now, I'm not familiar with the math behind ROSA and I'm not familiar with R, so I wasn't totally sure if this was conceptually correct.

Any insights would be appreciated!

**Comment by khliland** [comment]
Hi skylogic004,

Thank you for pointing out the bug regarding B0. I also found a missing parenthesis in rosa.fit(), which I cannot understand has not led to a warning previously. I have uploaded fixes to GitHub. Please test the updated if you have the opportunity.

Could you check if the ROSA examples work on your computer, i.e., run `example("rosa_results")`? In the example, the `predict()` function is used on a ROSA model. If these work as they should, it could also mean that there is something with the formatting of your data that the package does not recognize correctly. I have tested with various inputs to `predict()` without being able to reproduce the error you describe.

Regards,
Kristian

**Comment by skylogic004** [comment]
So, first, a side point: your definition of `ncomp` is sometimes an integer and sometimes a sequence - which makes things a bit confusing. e.g. in `predict.rosa` the default value is `1:object$ncomp`.

The problem isn't fixed yet. I'm calling predict like this:
```
predict(result, newdata=newdata)
```
which means I'm leaving `ncomp` and `comps` as their default values.
That means this code block is executed:
https://github.com/khliland/multiblock/blob/8b8a9236b07023c1c4c436870c3939f7254a66fb/R/rosa_results.R#L96
(so, the issue is in lines 96 to 109).

You've updated line 113, which is executed only when `comps` is provided, so that's not where my problem is.

Related to this, why is the prediction code in line 114 so much different than line 108? (shown below)
https://github.com/khliland/multiblock/blob/8b8a9236b07023c1c4c436870c3939f7254a66fb/R/rosa_results.R#L108
https://github.com/khliland/multiblock/blob/8b8a9236b07023c1c4c436870c3939f7254a66fb/R/rosa_results.R#L114

By the way, what does "cumulative coefficients" mean?

Thanks for looking into this,
Matt

**Comment by khliland** [comment]
The `ncomp` sometimes being a vector and sometimes an integer is unfortunate with regard to consistency, but too late to do anything about it. This package was first created without my involvement and is now used in too many other packages to change such basic stuff.

Regarding the error you are experiencing, I have not changed any code as I am unable to reproduce the error. I do not have a basis for doing changes. To be able to start looking for the reason for the error message, I would need access to an example where the error occurs. Do you get the same error if you use the dataset from the examples or only on your own data? I still think there could be something about formatting that the package doesn't understand.

Lines 108 and 114 are doing the same job, one in a loop, one in a single pass. The difference is in the lines before each of them where one uses models where all components up to the chosen one are used, i.e., the standard "cumulative" way, while the other can select a subset of components in a non-cumulative way.

Thanks for your feedback and hope you have more. :)
Regards,
Kristian

**Comment by skylogic004** [comment]
Hi Kristian, sorry I wasn't able to get to this sooner. Upon diving into this issue more, I've come to realize that there are two issues:

1. An issue with my data structure; how I was creating the matrix variables was causing an issue.

2. An issue with the predict function.

I've created a example with a toy dataset. There are two versions:
- [example_usingDF](https://gist.github.com/skylogic004/b15268d86aabe7356be967fd62e2fef4)
- [example_usingMatrix](https://gist.github.com/skylogic004/f22f92af408ce65f32a5deee9ca72784)

**For issue 1**: I created sample data in two different ways in `usingDF` and `usingMatrix`. 

`example_usingDF` works but `example_usingMatrix` produces the error I reported in my first post ("length of 'dimnames' [1] not equal to array extent"). 

The code in the two examples are the same, except for the way the data structure is created. I'm not familiar enough with R to know how/why these structures are treated differently.

**For issue 2**: Regardless of the data structure, both produce different predictions between the following calls, but I believe these calls should be identical? -->

`predict(result, comps=1:3)`
and
`predict(result, newdata=traindata, comps=1:3)`


And this is even more of a problem when passing in the testdata for newdata, because `predict(result, newdata=testdata, comps=1:3)` appears to make the wrong predictions. So I made my own predict function to correct it; it uses code copied from your library. All this is shown in the example code.

Hope this helps!
Matt

**Comment by khliland** [comment]
Hi Matt,

This was super helpful!

**Issue 1** is due to a side effect of using `cbind()` on the generated data instead of `data.frame()`, where the latter automatically adds `dimnames`, while the former does not. I have added checks for input data without `dimnames` in the ROSA and SO-PLS functions to catch and compensate for such cases. For future reference, there is also a `block.data.frame()` convenience function available in the package for constructing datasets.

**Issue 2** is hopefully fixed now. I had committed an update to GitHub after your previous descriptions which fixed part of the problems, but your excellent example highlighted one case that I had not covered. With the current update, all cases seem to run satisfactorily. 

Thank you very much for your help!
Kristian
